src/Symfony/Bridge/Twig/Debug/TwigFlattenExceptionProcessor.php

```diff
@@ -0,0 +1,140 @@
+class TwigFlattenExceptionProcessor implements FlattenExceptionProcessorInterface
+{
+    private $files = array();
+    private $twig;
+    private $loadedTemplates;
+
+    public function __construct(\Twig_Environment $twig)
+    {
+        $this->twig = $twig;
+        $this->loadedTemplates = new \ReflectionProperty($twig, 'loadedTemplates');
+        $this->loadedTemplates->setAccessible(true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(\Exception $exception, FlattenException $flattenException, $master)
+    {
+        $trace = $flattenException->getTrace();
+        $origTrace = $exception->getTrace();
+
+        foreach ($origTrace as $key => $entry) {
+            $prevKey = $key - 1;
+
+            if (!isset($origTrace[$prevKey]) || !isset($entry['class']) || 'Twig_Template' === $entry['class'] || !is_subclass_of($entry['class'], 'Twig_Template')) {
+                continue;
+            }
+
+            $template = $this->getLoadedTemplate($entry['class']);
+
+            if (!$template instanceof \Twig_Template) {
+                continue;
+            }
+
+            $file = $this->findOrCreateFile($template);
+
+            $data = array('file' => $file);
+            if (isset($origTrace[$prevKey]['line'])) {
+                $data['line'] = $this->findLineInTemplate($origTrace[$prevKey]['line'], $template);
+            }
+
+            $trace[$prevKey]['related_codes'][] = $data;
+        }
+
+        if (isset($trace[-1]) && $exception instanceof \Twig_Error) {
+            $name = $exception->getTemplateFile();
+            $file = $this->findOrCreateFile($name, $name);
+
+            $trace[-1]['related_codes'][] = array('file' => $file, 'line' => $exception->getTemplateLine());
+        }
+
+        $flattenException->replaceTrace($trace);
+    }
+
+    private function getLoadedTemplate($class)
+    {
+        $loadedTemplates = $this->loadedTemplates->getValue($this->twig);
+
+        return isset($loadedTemplates[$class]) ? $loadedTemplates[$class] : null;
+    }
+
+    private function findOrCreateFile($template, $path = null)
+    {
+        $name = $template instanceof \Twig_Template ? $template->getTemplateName() : $template;
+
+        if (isset($this->files[$name])) {
+            return $this->files[$name];
+        }
+
+        foreach ($this->files as $key => $file) {
+            if (isset($file->path) && $file->path == $name) {
+                return $file;
+            }
+        }
+
+        $file = (object) array('name' => $name, 'type' => 'twig');
+
+        try {
+            $path = $path ?: $this->twig->getLoader()->getCacheKey($name);
+        } catch (\Twig_Error_Loader $e) {
+        }
+
+        if (is_file($path)) {
+            $file->path = $path;
+        } else {
+            $source = null;
+
+            if (method_exists($template, 'getSource')) {
+                $source = $template->getSource();
+            }
+
+            if (null === $source) {
```

Comment:
I'd move this `if` as the `else` of the previous one (and remove `$source = null` initialization)